modules/nf-core/custom/matrixfilter/templates/matrixfilter.R

```diff
@@ -152,17 +154,42 @@ if ((opt\$sample_file != '') && ( ! is.null(opt\$grouping_variable))){
     opt\$minimum_samples <- ncol(abundance_matrix) * opt\$minimum_proportion
 }
 
+# Also set up filtering for NAs; use by default minimum_proportion_not_na; only
+# use minimum_samples_not_na if it is provided (default NULL)
+prefix = ifelse('$task.ext.prefix' == 'null', '', '$task.ext.prefix')
+
+if (is.null(opt\$minimum_samples_not_na)) {
+    opt\$minimum_samples_not_na <- ncol(abundance_matrix) * opt\$minimum_proportion_not_na
+}
+
+# Prepare variables that are needed in the apply (prefix is also needed later)
+
+prefix = ifelse('$task.ext.prefix' == 'null', '', '$task.ext.prefix')
+rowcounter <- 1 # This keeps track of the current row to allow rowname extraction
+
 # Generate a boolean vector specifying the features to retain
 
 keep <- apply(abundance_matrix, 1, function(x){
-    sum(x > opt\$minimum_abundance) >= opt\$minimum_samples
+    # Check if all or enough entries in the current row have a value
+    na_test <- !any(is.na(x)) || sum(!is.na(x))/length(x) >= opt\$minimum_samples_not_na
+
+    # Check if there is a high enough abundance in the current row
+    sum_test <- sum(x > opt\$minimum_abundance, na.rm = T) >= opt\$minimum_samples
+
+    # Log feature IDs that fail a test
+    if (!na_test) {
+        write(rownames(abundance_matrix)[rowcounter], file=paste0(prefix, '.NA_removed.txt'), append=T)
+    }
+    if (!sum_test) {
+        write(rownames(abundance_matrix)[rowcounter], file=paste0(prefix, '.abundance_removed.txt'), append=T)
+    }    
+    rowcounter <<- rowcounter+1
+    na_test && sum_test
```

Comment:
```suggestion
    sum(x > opt\$minimum_abundance, na.rm = TRUE) >= opt\$minimum_samples
})

keep_na <- apply(abundance_matrix, 1, function(x){
    !any(is.na(x)) || sum(!is.na(x))/length(x) >= opt\$minimum_samples_not_na
```

Suggest you do this - create two separate boolean vectors. Your final boolean vector will then be `keep & keep_na` (or whatever you'd like to call it).

You can use these vectors to select out the row IDs that failed for the different reasons.

I also suggest a single output file for the rejected rows, with row ids and 'reason' or somesuch, so that we can extend this in future without an expanding list of files-per-reason.

Reply:
Done, the output table now looks like this:

| Feature ID | Reason for rejection |
| ---------- | -------------------- |
| E9PVP1;A0A0G2JG52 | Abundance test |
| Q9D773 | Abundance test |
| Q9QYC3 | NA test |
| Q9QZQ8 | NA test |
| Q9Z2U4 | NA test, Abundance test |
| Q9Z2V6 | NA test, Abundance test |